src/python/zquantum/core/interfaces/ansatz.py

```diff
@@ -0,0 +1,79 @@
+from abc import ABC, abstractmethod
+from ..circuit import Circuit
+from .ansatz_utils import ansatz_property
+from ..utils import create_symbols_map
+import copy
+import sympy
+from typing import List, Optional
+from overrides import EnforceOverrides
+import numpy as np
+
+
+class Ansatz(ABC, EnforceOverrides):
+
+    supports_parametrized_circuits = None
+    n_layers = ansatz_property("n_layers")
+
+    def __init__(self, n_layers: int):
+        """
+        Interface for implementing different ansatzes.
+        This class also caches the circuit for given ansatz parameters.
+
+        Args:
+            n_layers: number of layers of the ansatz
+        
+        Attributes:
+            n_layers (int): see Args
+            circuit (zquantum.core.circuit.Circuit): circuit representation of the ansatz.
+            supports_parametrized_circuits(bool): flag indicating whether given ansatz supports parametrized circuits.
+        
+        """
+        self.n_layers = n_layers
+        self._parametrized_circuit = None
+
+    @property
+    def parametrized_circuit(self) -> Circuit:
+        if self._parametrized_circuit is None:
+            if self.supports_parametrized_circuits:
+                return self._generate_circuit()
+            else:
+                raise (
+                    Exception(
```

Comment:
`Exception` is a bit vague. Not sure what's better though. Perhaps: `NotImplementedError`, `TypeError`, or `RuntimeError`. Maybe @dexter2206 has a better idea.

Reply:
I recommend `NotImplementedError`.

Reply:
Adressed in [9b2fca](https://github.com/zapatacomputing/z-quantum-core/pull/42/commits/9b2fcafb8de5209328bae0582eaaab2a8a0a374b).